[Alson Ang] iP

alson001 opened the pull request:
#Duke 
> "A work-room should be like an old shoe; no matter how shabby, it's better than a new one." - Author: Willa Cather

Duke list and record your deadlines, events and todos so that you can keep track of your tasks. It's
- Simple to use
- text-based
- ~~Fast~~ _Super_  Fast
All you need to do is to:

1. Download from [here](https://github.com/alson001/ip/releases/tag/A-Jar)
2. Run java -jar on the command line
3. Add your tasks.
4. Let it manage your tasks for you 👍 

And it's **FREE**

Features:
- [x] Todo
- [x] Deadline 
- [x] Event 
- [x] Find
- [x] bye, list, mark

If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```ruby
public static void main(String[] args) {
       Duke duke = new Duke();
       duke.run();
}
```

Commit:
Add Gradle support

alson001 committed:
Level 1. Greet, Echo, Exit

alson001 committed:
Level 2. Add, List

alson001 committed:
Level 3. Mark as Done

alson001 committed:
Level 4. ToDos, Events, Deadlines

alson001 committed:
A-TextUiTesting: Automated Text UI Testing

alson001 committed:
Level 5. Handle Errors

alson001 committed:
Level 6. Delete

alson001 committed:
A-Enums: Use Enums

alson001 committed:
Add saving and loading of tasks to hard disk (Level-7)

alson001 committed:
Add Dates and Times (Level 8)

alson001 committed:
Merge branch 'branch-Level-8'

alson001 committed:
Add more OOP

alson001 committed:
Organize into Packages

alson001 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

alson001 committed:
Add gradle

alson001 committed:
Add JUnit tests

alson001 committed:
Add JavaDoc

alson001 committed:
Tweak code to comply with coding standard

alson001 committed:
Add find command

alson001 committed:
Merge branch 'branch-A-CodingStandard'

alson001 committed:
Merge branch 'branch-Level-9'

alson001 committed:
Use CheckStyle

edures888 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,98 @@
+    /**
+     * Prints a bye message to the user.
+     */
+    public void byeMessage() {
```
Maybe it would be better to use verbs for the methods?

alson001 replied:
I think this is good enough.

edures888 commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,127 @@
+     * @param userCommands The string command to parse.
+     * @param commandList The task list that has list of tasks.
+     * @param storage The storage that store tasks from previous and new sessions.
+     * @param ui The ui used for the bot.
+     * @param file The file that data is read and written from.
+     * @return The command corresponding to the user input.
+     * @throws InvalidCmdValueException If a delete, mark or unmark command specify a wrong index.
+     * @throws InvalidTaskTypeException If no such task type exists.
+     * @throws EmptyCommandException If the existing command is empty
+     * @throws InvalidTimeException If the format of event or deadline is unrecognised.
+     * @throws InvalidDateException If the date entered for event or deadline is unrecognised.
+     */
+
+    public Command parse(String userCommands, TaskList commandList,
+                         Storage storage, Ui ui, File file)
+            throws InvalidCmdValueException, InvalidTaskTypeException,
+            EmptyCommandException, InvalidTimeException, InvalidDateException {
+        String[] strArray = userCommands.split(" ", 2);
+        String action = strArray[0];
+
+        if (action.equalsIgnoreCase("bye")) {
+            return new ByeCommand(ui);
+        } else if (action.equalsIgnoreCase("list")) {
+            return new ListCommand(ui, commandList);
+        } else if (action.equalsIgnoreCase("mark")) {
+            return new MarkCommand(ui, commandList,
+                    Integer.parseInt(strArray[1]) - 1, storage, file);
+        } else if (action.equalsIgnoreCase("unmark")) {
+            return new UnmarkCommand(ui, commandList,
+                    Integer.parseInt(strArray[1]) - 1, storage, file);
+        } else if (action.equalsIgnoreCase("delete")) {
+            return new DeleteCommand(ui, commandList,
+                    Integer.parseInt(strArray[1]) - 1, storage, file);
+        } else if (action.equalsIgnoreCase("find")) {
```
I think it would be better to use named constants instead of magic strings.

alson001 replied:
I think there might be error if I declare
String index = strArray[1];
when the person only inputs one word as strArray[1] will not exist.

edures888 commented on src/main/java/duke/exceptions/InvalidCmdValueException.java:
```diff
@@ -0,0 +1,8 @@
+package duke.exceptions;
+
+public class InvalidCmdValueException extends Exception {
```
I like the level of abstraction for exceptions, but maybe spelling out the full word would be better for naming?

alson001 replied:
Renamed with full spelling.

edures888 commented on src/main/java/duke/tasks/TaskTypes.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * TaskTypes represents the different types of Task.
+ */
+public enum TaskTypes {
```
I like the usage of enumerations here for the types of tasks!

edures888 reviewed (commented):
Overall, I think there's great progress! Just a few nits here and there to fix.

alson001 committed:
Add a GUI to Duke

alson001 committed:
Merge branch 'branch-level-10'

cmang12 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,74 @@
+    private Parser parser;
+
+    @Override
+    public void start(Stage stage) throws Exception {
```
Perhaps you can have a Main class to control Duke and the MainWindow?

alson001 replied:
I like it combined.

cmang12 commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * AddCommand represents a command to add a task to the TaskList.
+ */
+public class AddCommand extends Command {
```
I like that you have different classes for different commands, very OOP :D

cmang12 commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,48 @@
+package duke.commands;
```
I like the way you used packages

cmang12 commented on src/main/java/duke/exceptions/EmptyCommandException.java:
```diff
@@ -0,0 +1,16 @@
+ * EmptyCommandException represents an exception if
+ * the existing command is empty.
+ */
+public class EmptyCommandException extends Exception {
```
I like that the names of your exceptions are clear and appropriate for each error.

cmang12 commented on src/main/java/duke/tasks/TaskTypes.java:
```diff
@@ -0,0 +1,13 @@
+/**
+ * TaskTypes represents the different types of Task.
+ */
+public enum TaskTypes {
```
I like that you are using enums!

cmang12 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,101 @@
+     * Prints a response to Find command to the user.
+     * @param taskList The filtered task list
+     */
+    public String findResponse(List<Task> taskList) {
```
I like the method names here, v clear

cmang12 reviewed (commented):
Overall, I found your code easy to read. I also really liked that you adhered to OOP principles. Keep up the good work 💯

hhchinh2002 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,74 @@
+            return response;
+        } catch (InvalidCmdValueException | InvalidTaskTypeException
+                 | EmptyCommandException | InvalidTimeException | InvalidDateException e) {
+            return Ui.HORIZONTAL_LINE + "\n" + e.getMessage() + "\n" + Ui.HORIZONTAL_LINE;
```
Would it be neater if we have a method (e.g from Ui class) to put the message in wanted form (in this case, it's placed between2 horizontal lines) and calling it whenever we want to print something to the user?

alson001 replied:
Made a `formatString()` method to UI.

hhchinh2002 commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ *  A launcher for the application.
+ */
+public class Launcher {
```
I like this practice of making a new class representing for the launcher

hhchinh2002 commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,67 @@
+        dialogContainer.getChildren().addAll(
+                DialogBox.getUserDialog(input, userImage),
+                DialogBox.getDukeDialog(response, dukeImage)
+        );
```
Would putting the ")" when invoking to a method in a newline here violates our coding standards?

alson001 replied:
Removed newline on line63.

hhchinh2002 commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,48 @@
+
+    /**
+     * Creates a AddCommand to add a ToDo, Event or Deadline to the TaskList.
+     * @param ui The ui used.
```
Should there be an empty line between the method description and parameters section?

alson001 replied:
Yes, there should be. Inserted an empty line between the method description and parameters section for all comments.

hhchinh2002 commented on src/main/java/duke/commands/ByeCommand.java:
```diff
@@ -0,0 +1,28 @@
+    private Ui ui;
+
+    /**
+     * Creates a ByeCommand to say bye.
```
Same suggestion as the previous comment

hhchinh2002 commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,54 @@
+    private File file;
+
+    /**
+     * Creates a DeleteCommand to delete a ToDo, Event or Deadline from the TaskList.
```
Same suggestions as previous comments

hhchinh2002 commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,41 @@
+    /**
+     * Creates a FindCommand to find all tasks on the Task list
+     * based on similarity of a given string.
+     * @param ui The ui used.
```
Same suggestion as previous comments, there are other places with the same behaviour too.

hhchinh2002 reviewed (commented):
Overall, I think the code are almost perfect for me, I really like your way of OOP-izing everything as much as possible.

alson001 committed:
Arrange saved data

alson001 committed:
Add assertions in Command classes

DeleteCommand updates a Task. TaskList marks/unmarks a Task.
FindCommand queries for a List of Tasks.

There may be a possibility that the program fails to perform these
actions and return null.

Assertions are added to ensure these are done successfully.

Assertions added to following classes:
* TaskList
* FindCommand
* DeleteCommand

alson001 committed:
Improve code quality

Previously, for JavaDocs, the method description and parameters section
lacks a newline. Rename InvalidCmdValueException to
InvalidCommandValueException for clearer readability. New method
formatString() in UI that formats string for better OOP.
DialogBox has been edited to fit information better.

alson001 committed:
Merge pull request #1 from alson001/branch-A-Assertions

Add assertions in Command classes

alson001 committed:
Merge pull request #2 from alson001/branch-A-CodeQuality

Improve code quality

alson001 committed:
Add undo command.

Calling undo multiple times will undo previous commands in the current
session on a FIFO basis. If there is nothing to undo, an exception will
be thrown.

Commands that can be undone includes Add, Delete, Mark, Unmark
while commands that cannot are List, Find, Exit, Undo.

alson001 committed:
Merge pull request #3 from alson001/branch-C-Undo

Add undo command.

alson001 committed:
Add User Guide

alson001 committed:
Fix issue with build

alson001 committed:
Fix `list` bug when there is an empty list

alson001 committed:
Fix empty commands for `mark`, `unmark` and `delete`

alson001 committed:
Rename Ui.png

alson001 committed:
Fix error where if savedfile.txt is missing, application will not
launch
